scipy/interpolate/src/__fitpack.cc

```diff
@@ -0,0 +1,426 @@
+        }
+        offset(i) = ind - k;
+
+        // compute non-zero b-splines
+        _deBoor_D(t.data, xval, k, ind, 0, wrk);
+
+        for (ssize_t j=0; j < k+1; ++j) {
+            A(i, j) = wrk[j] * w(i);
+        }
+    }
+
+    *nc = len_t - k - 1;
+}
+
+
+/*
+ *   Solve the LSQ problem ||y - A@c||^2 via QR factorization.
+ *  
+    QR factorization follows FITPACK: we reduce A row-by-row by Givens rotations.
+    To zero out the lower triangle, we use in the row `i` and column `j < i`,
+    the diagonal element in that column. That way, the sequence is
+    (here `[x]` are the pair of elements to Givens-rotate)
+
+     [x] x x x       x  x  x x      x  x  x x      x x  x  x      x x x x
+     [x] x x x  ->   0 [x] x x  ->  0 [x] x x  ->  0 x  x  x  ->  0 x x x
+      0  x x x       0 [x] x x      0  0  x x      0 0 [x] x      0 0 x x
+      0  x x x       0  x  x x      0 [x] x x      0 0 [x] x      0 0 0 x
+
+    The matrix A has a special structure: each row has at most (k+1) non-zeros, so
+    is encoded as a PackedMatrix instance.
+
+    On exit, the return matrix, also of shape (m, k+1), contains
+    elements of the upper triangular matrix `R[i, i: i + k + 1]`.
+    When we process the element (i, j), we store the rotated row in R[i, :],
+    and *shift it to the left*, so that the the diagonal element is always in the
+    zero-th place. This way, the process above becomes
+
+     [x] x x x       x  x x x       x  x x x       x  x x x      x x x x
+     [x] x x x  ->  [x] x x -  ->  [x] x x -   ->  x  x x -  ->  x x x -
+      x  x x -      [x] x x -       x  x - -      [x] x - -      x x - -
+      x  x x -       x  x x -      [x] x x -      [x] x - -      x - - -
+
+    The most confusing part is that when rotating the row `i` with a row `j`
+    above it, the offsets differ: for the upper row  `j`, `R[j, 0]` is the diagonal
+    element, while for the row `i`, `R[i, 0]` is the element being annihilated.
+
+    NB. This row-by-row Givens reduction process follows FITPACK:
+    https://github.com/scipy/scipy/blob/maintenance/1.11.x/scipy/interpolate/fitpack/fpcurf.f#L112-L161
+    A possibly more efficient way could be to note that all data points which
+    lie between two knots all have the same offset: if `t[i] < x_1 .... x_s < t[i+1]`,
+    the `s-1` corresponding rows form an `(s-1, k+1)`-sized "block".
+    Then a blocked QR implementation could look like
+    https://people.sc.fsu.edu/~jburkardt/f77_src/band_qr/band_qr.f
+
+    The `startrow` optional argument accounts for the scenatio with a two-step
+    factorization. Namely, the preceding rows are assumend to be already
+    processed and are skipped.
+    This is to account for the scenario where we append new rows to an already
+    triangularized matrix.
+
+    This routine MODIFIES `a` & `y` in-place.
+
+ */
+void
+qr_reduce(double *aptr, const ssize_t m, const ssize_t nz, // a(m, nz), packed
```

Comment:
Typically Householder is used for this via [`dlarfg`](https://www.netlib.org/lapack/explore-html/d8/d0d/group__larfg_gadc154fac2a92ae4c7405169a9d1f5ae9.html#gadc154fac2a92ae4c7405169a9d1f5ae9) instead of Givens but it is what it is from the Fortran code. It is OK I guess.

Reply:
Yeah, I left a lengthy comment on how this can be changed into a blocked Householder QR, https://github.com/scipy/scipy/pull/19970/files#diff-8da5866d31e220e1e6d3853b9c7f262a1a034ac88f51f0fc47dcad511397b8a0R201
For the time being, I'd prefer to keep the row-by-row Givens process to follow what FITPACK was doing (and to avoid a ton of bookkeeping, too :-)).

Reply:
It's actually simpler with householder as it holds the reflectors in the same array but probably going rogue in the middle of the original code is not a good move anyways.